torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -223,6 +223,22 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
                 tensor_meta=self._tp_spec.tensor_meta,
             )
             param_data = cast(DTensor, param)._local_tensor
+
```

Comment:
@awgu Want to check with you whether to disable the 2D state_dict like this is ok with you. 

Originally thinking about doing it in `_register_state_dict_hooks`, but since `_register_state_dict_hooks` happens during lazy_init so ended up registering the hooks in param init.  
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_composable/fsdp/_fsdp_param_group.py#L469

Reply:
I think we can move this to the end of `FSDPParamGroup.__init__` so that we do not need the `if len(cur_module._state_dict_pre_hooks) == 0:` check in the case that a module has >1 parameter.

Reply:
I was looking to add this in `FSDPParamGroup` initially, but since we only wanted to add this hook for 2D, then we would need an `if` in `FSDPParamGroup.__init__` as well. Something like this:

```
if any(len(fsdp_param._spmd_placements)>1 for fsdp_param in self.fsdp_params):
    self.module. register_state_dict_pre_hook(...)
    self.module. _register_load_state_dict_pre_hook(...)
```
In this case, we would also run the extra if check for 1D vs. We only register the hook if we encounter a 2D placements so nothing changes from 1D side, but we will do the `_state_dict_pre_hooks` check for every param init for 2D. 

Not sure which one is more preferrable to you.

Reply:
Discussed with @awgu offline to move the check to the end of FSDPParamGroup.__init__.